adr/ADR-40.md

```diff
@@ -125,6 +126,46 @@ all reconnect options are respected.
 For most clients, that means having a `reconnect` method on the
 Client/Connection handle.
 
+#### Custom server pool
+
+Client should have a way to specify custom server pool for reconnection
+attempts. When specified, client should use only those servers for reconnection
+attempts, ignoring any advertised servers from the Server.
+
+New pool should be used for all reconnection attempts until the client is
+restarted or a new pool is specified. New pool is subject to the same rules as
+the default one - randomization unless disabled, max reconnect attempts, etc.
+
+```go
+// SetServerPool sets a new server pool for reconnection attempts.
+// It replaces any existing pool, including advertised servers.
+func (nc *Conn) SetServerPool(servers []string) error
+```
+
+#### Reconnect to specific server
+
+Client should expose an option to configure a callback that will be called
+before each reconnection attempt.
+
+As arguments, it should receive the list of servers that are in the pool
+(together with reconnect attempt number), as well as the most up-to-date server
+INFO.
```

Comment:
not quite sure why we want to expose all the metrics - the callbacks for reconnecting/reconnect/disconnect have info on what is happening, and grabbing the info for their investigation is available on connection - this means that they can grab/store/use their own data to supplement the selection the "server selector".

Effectively, I would move my internal selection logic to such function and use it internally, this way how it's handled is uniform.